client/a8c-for-agencies/components/a4a-feedback/churn-mechanism/cancel-license-feedback-modal.tsx

```diff
@@ -0,0 +1,333 @@
+} ) => {
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+
+	const siteSlug = siteUrl ? urlToSlug( siteUrl ) : null;
+
+	const refreshLicenceList = useRefetchLicenses( LicensesOverviewContext );
+
+	const [ comments, setComments ] = useState< string >( '' );
+	const [ suggestions, setSuggestions ] = useState< FeedbackSuggestion[] >( [] );
+	const [ selectedCTA, setSelectedCTA ] = useState< 'cancel' | 'speak-with-partner-manager' >();
+
+	const onSuggestionChange = ( option: FeedbackSuggestion ) => {
+		if ( suggestions.find( ( suggestion ) => suggestion.value === option.value ) ) {
+			setSuggestions( ( prev ) =>
+				prev.filter( ( suggestion ) => suggestion.value !== option.value )
+			);
+		} else {
+			setSuggestions( ( prev ) => [ ...prev, option ] );
+		}
+	};
+
+	const { suggestion } = getA4AfeedbackProps( FeedbackType.LicenseCancelProduct, translate, {
+		productName,
+	} );
+
+	const isPressableLicense = isPressableHostingProduct( licenseKey );
+	const isWPCOMHostingLicense = isWPCOMHostingProduct( licenseKey );
+
+	const isHostingLicense = isPressableLicense || isWPCOMHostingLicense;
+
+	const type: FeedbackType = isHostingLicense
+		? FeedbackType.LicenseCancelHosting
+		: FeedbackType.LicenseCancelProduct;
+
+	const agencyId = useSelector( getActiveAgencyId );
+
+	const {
+		mutate: saveFeedback,
+		status: saveFeedbackStatus,
+		isPending: isSavingFeedback,
+	} = useSaveFeedbackMutation();
+
+	const {
+		mutate: revokeLicense,
+		isPending: isRevokingLicense,
+		status: revokeLicenseStatus,
+	} = useRevokeLicenseMutation();
+
+	const isLoading = isRevokingLicense || isSavingFeedback;
+
+	// If the user clicks cancel and provides feedback, the feedback is valid
+	// If the user clicks cancel and does not provide feedback, the feedback is not valid
+	// If the user clicks speak with partner manager and provides feedback, the feedback is valid
+	// If the user clicks speak with partner manager and does not provide feedback, the feedback is valid
+	const isFeedbackValid = useCallback(
+		( cta: 'cancel' | 'speak-with-partner-manager' ) => {
+			if ( cta === 'cancel' ) {
+				return comments || suggestions.length;
+			}
+			return true;
+		},
+		[ comments, suggestions ]
+	);
+
+	const handleSubmitFeedback = ( cta: 'cancel' | 'speak-with-partner-manager' ) => {
+		if ( agencyId && isFeedbackValid( cta ) ) {
+			const params = {
+				site_id: agencyId,
+				survey_id: type,
+				survey_responses: {
+					comment: { text: comments },
+					suggestions: {
+						text: suggestions?.map( ( suggestion ) => suggestion.value ).join( ', ' ) || '',
+					},
+					cta,
+				},
+			};
+			dispatch(
+				recordTracksEvent( 'calypso_a4a_churn_feedback_submit', {
+					agency_id: agencyId,
+					survey_id: params.survey_id,
+					comment: params.survey_responses.comment.text,
+					suggestions: params.survey_responses.suggestions.text,
+					cta: params.survey_responses.cta,
+				} )
+			);
+
+			saveFeedback( { params } );
```

Comment:
Regarding my comment about the `useEffect` below, can we place the post-success logic here to make it easier to follow?

```

			saveFeedback(
				{ params },
				{
					onSuccess: () => {
						// DO success here.
					},
				}
			);
```

Reply:
Replied in the above conversation.